src/back/hlsl/writer.rs

```diff
@@ -2289,11 +2292,9 @@ impl<'a, W: fmt::Write> super::Writer<'a, W> {
                     crate::VectorSize::Tri => "xxx",
                     crate::VectorSize::Quad => "xxxx",
                 };
-                let resolved = func_ctx.info[expr].ty.inner_with(&module.types);
-                self.write_value_type(module, resolved)?;
                 write!(self.out, "(")?;
                 self.write_expr(module, value, func_ctx)?;
-                write!(self.out, ".{})", number_of_components)?
+                write!(self.out, ").{}", number_of_components)?
```

Comment:
This is definitely wrong, the parentheses is there to stop problems with operator precedence so it needs to be outside.

Reply:
This is for `Expression::Splat`; the outside parentheses were only needed because the splat was previously wrapped in a constructor which I don't think is needed. Validation failed here for boolean splat ([see CI](https://github.com/gfx-rs/naga/runs/5991330322?check_suite_focus=true)); surrounding the boolean and then writing the splat components fixes it.

Reply:
wow ok, I don't work much with hlsl so I wasn't aware of this stuff, but still if you don't use the constructor anymore why is the parentheses there? Isn't the hlsl backend generating the code having in attention the expression precedence? It seems to me like it especially when you look at `Expression::Swizzle` which is basically what your are outputting but without parentheses.

Reply:
Filed https://github.com/microsoft/DirectXShaderCompiler/issues/4393

Reply:
I think we should leave a comment here about filled issue. Just to make sure that we didn't confuse other contributors.

Reply:
Ok makes sense to me